util/stmtsummary/evicted.go

```diff
@@ -0,0 +1,158 @@
+package stmtsummary
+
+import (
+	"container/list"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// stmtSummaryByDigestEvicted contents digests evicted from stmtSummaryByDigestMap
+type stmtSummaryByDigestEvicted struct {
+	// record evicted data in intervals
+	// latest history data is Back()
+	history *list.List
+}
+
+// element being stored in stmtSummaryByDigestEvicted
+type stmtSummaryByDigestEvictedElement struct {
+	// beginTime is the begin time of current interval
+	beginTime int64
+	// endTime is the end time of current interval
+	endTime int64
+	// *Kinds* of digest being evicted
+	digestKeyMap map[string]struct{}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvicted
+func newStmtSummaryByDigestEvicted() *stmtSummaryByDigestEvicted {
+	return &stmtSummaryByDigestEvicted{
+		history: list.New(),
+	}
+}
+
+// spawn a new pointer to stmtSummaryByDigestEvictedElement
+func newStmtSummaryByDigestEvictedElement(beginTime int64, endTime int64) *stmtSummaryByDigestEvictedElement {
+	return &stmtSummaryByDigestEvictedElement{
+		beginTime:    beginTime,
+		endTime:      endTime,
+		digestKeyMap: make(map[string]struct{}),
+	}
+}
+
+// AddEvicted is used add an evicted record to stmtSummaryByDigestEvicted
+func (ssbde *stmtSummaryByDigestEvicted) AddEvicted(evictedKey *stmtSummaryByDigestKey, evictedValue *stmtSummaryByDigest, historySize int) {
```

Comment:
Add a unit test for this function. make sure more than 80% line can be covered by the unit test.

Reply:
My bad, I'll do it ASAP.

Reply:
I just write a tiny test for your reference only 

```go
func (s *testStmtSummarySuite) TestSimpleStmtSummaryByDigestEvicted(c *C) {
	ssbde := newStmtSummaryByDigestEvicted()

	evictedKey, evictedValue := s.newStmtSummaryByDigestKeyValue("a", 1, 2)
	ssbde.AddEvicted(evictedKey, evictedValue, 0)
	c.Assert(ssbde.history.Len(), Equals, 0)

	// TODO: add more case...
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	// check the ssbde.history value

	evictedKey, evictedValue = s.newStmtSummaryByDigestKeyValue("b", 2, 3)
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	// check the ssbde.history value

	evictedKey, evictedValue = s.newStmtSummaryByDigestKeyValue("c", 6, 7)
	ssbde.AddEvicted(evictedKey, evictedValue, 2)
	// check the ssbde.history value

	// and so on
}

func (s *testStmtSummarySuite) newStmtSummaryByDigestKeyValue(schema string, beginTime int64, endTime int64) (*stmtSummaryByDigestKey, *stmtSummaryByDigest) {
	key := &stmtSummaryByDigestKey{
		schemaName: schema,
	}
	value := newInduceSsbd(beginTime, endTime)
	return key, value
}

```

Reply:
Please add the following test and fix related case. Also better to add more test.

```go

func (s *testStmtSummarySuite) TestSimpleStmtSummaryByDigestEvicted(c *C) {
	// TODO: fix me
	ssbde := newStmtSummaryByDigestEvicted()
	evictedKey, evictedValue := s.newStmtSummaryByDigestKeyValue("a", 1, 2)
	ssbde.AddEvicted(evictedKey, evictedValue, 0)
	//c.Assert(ssbde.history.Len(), Equals, 0)

	ssbde = newStmtSummaryByDigestEvicted()
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	c.Assert(getAllEvicted(ssbde), Equals, "{begin: 1, end: 2, count: 1}")

	evictedKey, evictedValue = s.newStmtSummaryByDigestKeyValue("b", 1, 2)
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	c.Assert(getAllEvicted(ssbde), Equals, "{begin: 1, end: 2, count: 2}")

	evictedKey, evictedValue = s.newStmtSummaryByDigestKeyValue("b", 5, 6)
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	c.Assert(getAllEvicted(ssbde), Equals, "{begin: 5, end: 6, count: 1}, {begin: 1, end: 2, count: 2}")

	// todo: fix me
	evictedKey, evictedValue = s.newStmtSummaryByDigestKeyValue("b", 3, 4)
	ssbde.AddEvicted(evictedKey, evictedValue, 1)
	c.Assert(getAllEvicted(ssbde), Equals, "{begin: 5, end: 6, count: 1},  {begin: 3, end: 4, count: 1}, {begin: 1, end: 2, count: 2}")
}

func (s *testStmtSummarySuite) newStmtSummaryByDigestKeyValue(schema string, beginTime int64, endTime int64) (*stmtSummaryByDigestKey, *stmtSummaryByDigest) {
	key := &stmtSummaryByDigestKey{
		schemaName: schema,
	}
	value := newInduceSsbd(beginTime, endTime)
	value.schemaName = schema
	return key, value
}

func getAllEvicted(ssdbe *stmtSummaryByDigestEvicted) string {
	buf := bytes.NewBuffer(nil)
	for e := ssdbe.history.Back(); e != nil; e = e.Prev() {
		if buf.Len() != 0 {
			buf.WriteString(", ")
		}
		val := e.Value.(*stmtSummaryByDigestEvictedElement)
		buf.WriteString(fmt.Sprintf("{begin: %v, end: %v, count: %v}", val.beginTime, val.endTime, len(val.digestKeyMap)))
	}
	return buf.String()
}
```